File: Algorithms/p290_Word_Pattern/p290_Word_Pattern.py
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        mapping = [None]*256
        str_seen = set()
        s_list = str.split()
        if len(pattern)!=len(s_list):
            return False
        for c,s in zip(pattern,s_list):
            c = ord(c)
            if mapping[c]!=s:
                if mapping[c]==None:
                    if s not in str_seen:
                        str_seen.add(s)
                    else:
                        return False
                    mapping[c] = s;
                else:
                    return False
        return True

File: Algorithms/p290_Word_Pattern/p290_Word_Pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        vector<string> mapping(256);
        unordered_set<string> str_seen;
        int pos1 = 0;
        int pos2 = 0;
        int pos2_start;
        while(1){
            while(pos2<str.size()&&str[pos2]==' ') ++pos2;
            if(pos2>=str.size()) break;
            if(pos1>=pattern.size()) break;
            pos2_start = pos2;
            while(pos2<str.size()&&str[pos2]!=' ') ++pos2;
            char c = pattern[pos1++];
            string s = str.substr(pos2_start,pos2-pos2_start);
            if(mapping[c]!=s){
                if(mapping[c].size()==0){
                    if(!str_seen.insert(s).second) return 0;
                    mapping[c] = s;
                }else{
                    return 0;
                }
            }
        }
        return pos1==pattern.size()&&pos2==str.size();
    }
};
